x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -108,144 +129,165 @@ protected void masterOperation(
-                                .settings(
-                                    Settings.builder()
-                                        .put(indexMetadata.getSettings())
-                                        .put(IndexMetadata.INDEX_ROLLUP_SOURCE_NAME.getKey(), sourceIndexName)
-                                        .put(IndexMetadata.INDEX_ROLLUP_SOURCE_UUID.getKey(), sourceIndexUuid)
-                                )
-                        )
+            // 2. Create rollup index
+            createRollupIndex(rollupIndexName, sourceIndexMetadata, mapping, request, ActionListener.wrap(createIndexResp -> {
+                if (createIndexResp.isAcknowledged()) {
+                    // 3. Rollup index created. Run rollup indexer
+                    RollupIndexerAction.Request rollupIndexerRequest = new RollupIndexerAction.Request(
+                        request,
+                        dimensionFieldCaps.keySet().toArray(new String[0]),
+                        metricFieldCaps.keySet().toArray(new String[0])
                     );
-                }
+                    rollupIndexerRequest.setParentTask(clusterService.localNode().getId(), task.getId());
 
-                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                    // index created
-                    // 3.
                     client.execute(RollupIndexerAction.INSTANCE, rollupIndexerRequest, ActionListener.wrap(indexerResp -> {
                         if (indexerResp.isCreated()) {
-                            // 4.
+                            // 4. Make rollup index read-only and set the correct number of replicas
+                            final Settings settings = Settings.builder()
+                                .put(IndexMetadata.SETTING_BLOCKS_WRITE, true)
+                                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, sourceIndexMetadata.getNumberOfReplicas())
+                                .build();
+                            UpdateSettingsRequest updateSettingsReq = new UpdateSettingsRequest(settings, rollupIndexName);
+                            updateSettingsReq.setParentTask(clusterService.localNode().getId(), task.getId());
                             client.admin().indices().updateSettings(updateSettingsReq, ActionListener.wrap(updateSettingsResponse -> {
                                 if (updateSettingsResponse.isAcknowledged()) {
-                                    // 5.
-                                    client.admin().indices().resizeIndex(resizeRequest, ActionListener.wrap(resizeResponse -> {
-                                        if (resizeResponse.isAcknowledged()) {
-                                            // 6.
-                                            publishMetadata(originalIndexName, tmpIndexName, rollupIndexName, listener);
+                                    // 5. Add rollup index to data stream and publish rollup metadata
```

Comment:
Not sure why we should wait for the replicas to recover. I was thinking that this can happen in parallel. Any concerns we may lose the primary shard while deleting the source index?

I am afraid that waiting for replicas to recover may take lots of time.